Troupon.Catalog.Api/Controllers/DealsController.cs

```diff
@@ -0,0 +1,77 @@
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading.Tasks;
+using AutoMapper;
+using Infra.Api;
+using Infra.Api.Conventions;
+using MediatR;
+using Microsoft.AspNetCore.Mvc;
+using Microsoft.AspNetCore.Mvc.ModelBinding;
+using Swashbuckle.AspNetCore.Annotations;
+using Troupon.Catalog.Core.Domain.Dtos;
+using Troupon.Catalog.Core.Domain.InputModels;
+
+namespace Troupon.Catalog.Api.Controllers
+{
+  [ApiController]
+  [Route("api/v{version:apiVersion}/[controller]")]
+  [ApiVersion("1.0")]
+  [ApiVersion("2.0")]
+  [ApiConventionType(typeof(PwcApiConventions))]
+  public class DealsController : PwcBaseController
+  {
+    public DealsController(IMediator mediator, IMapper mapper)
+        : base(mediator, mapper)
+    {
+    }
+
+    [SwaggerOperation(
+      Description = "Returns n filtered deals",
+      OperationId = "deals")]
+    [HttpGet]
+    public Task<IEnumerable<DealDto>> Get([FromQuery] DealsSearchFilter value)
+    {
+      return Task.FromResult(
+        Enumerable.Repeat(new DealDto(), 20));
+    }
+
+    [SwaggerOperation(
+      Description = "Get deal details by id",
+      OperationId = "deals")]
+    [HttpGet("{id}")]
+    public Task<DealDetailsDto> GetById([BindRequired] Guid id)
+    {
+      return Task.FromResult(new DealDetailsDto());
+    }
+
+    [SwaggerOperation(
+     Description = "Get deal details by id",
+     OperationId = "deals",
+     Tags = new[] { "Example for v2 api" })]
+    [MapToApiVersion("2.0")]
+    [HttpGet("{id}")]
+    public Task<DealDetailsDto> GetByIdV2([BindRequired] Guid id)
+    {
+      return Task.FromResult(new DealDetailsDto());
+    }
+
+    [SwaggerOperation(
+     Description = "Returns the popular deals for home page",
+     OperationId = "popular-deals")]
+    [HttpGet("popular-deals")]
+    public Task<IEnumerable<DealDto>> Get()
```

Comment:
It's only a feeling, but should we have add endpoints for this ? Or it could be only something we add to DealsSearchFilter ?

Like
api/v1/deals => all
api/v1/deals?view=popular-deals
api/v1/deals?view=recent-search

If we want to stay "as close as possible" to REST ?

Reply:
I think that this approach is the REST friendly approach. Following your idea, we would have the same resource (get deal) handling three different actions (get all, get popular deals, get recent search) ? I'm not familiar with this approach in REST. Do we have any reference to this approach (using the same resource to different responsibilities)?

Reply:
I think you're right ! It could stay like that then !

I was sure that your approach was more endpoints then RESTful. But it's interesting because while searching I found good practices with REST.

1. Use nouns instead of verbs (ex: /deals/popular-deals is OK ; /orders/place-order NOT)
2. Have a path consistency (ex: /{ressource}/{id}/{sub-ressource}) the structure is not important, but should be as standardize as possible.
3. Make endpoints stateless (no session variables or other "cookie" mechanisim, etc)
4. Use HTTP Methods the right way (GET retrives ressources; POST submit new data; PUT updates existing data; DELETE removes data)
5. Use logical nesting, could be nice (ex: GET /deals/{id}/options/ instead of /deal-options/{deal-id})
6. And many more hehe ...

My feeling here what more that it's like if "popular-deals" it's the view id of a filter or somethings. So it's more a question: Should we want to standardize our paths now or we prefer to wait later to see what shape it will take ?

like: /{ressource}/{id}/{sub-ressource}/{sub-id}


If yes:
We could say by example, when we have a special view for a list of something, we want to do.

/{ressource}/all/views/{view-name}
ex:
/deals/all/views/popular-deals

For real, you can totally merge like that !! 💯 I see with others to think about it later.

What do you thinks ?


some references:
- https://restfulapi.net/resource-naming/
- https://stackoverflow.blog/2020/03/02/best-practices-for-rest-api-design/
- https://searchapparchitecture.techtarget.com/definition/RESTful-API

Reply:
Sounds great! Let's try to discuss this on the meeting that we are going to have this afternoon